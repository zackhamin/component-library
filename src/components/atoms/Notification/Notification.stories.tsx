import { colors } from 'Styles';
import { Notification, NotificationProps } from './Notification';

export default { title: 'Atoms/Notification' };
export const NotificationStory = (args: NotificationProps) => (
	<Notification {...args} />
);

NotificationStory.storyName = 'Notification';
NotificationStory.args = {
	children: 'Customer has been updated successfully',
	color: colors.emerald,
	backgroundColor: colors.white,
	hasBorder: true,
	isDismissible: true,
};
