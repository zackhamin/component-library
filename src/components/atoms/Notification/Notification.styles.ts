import { NotificationProps } from './Notification';
import styled from 'styled-components';

export const StyledNotification = styled.div<Partial<NotificationProps>>`
	background: ${({ backgroundColor }) => backgroundColor};
	border-left: ${({ hasBorder, color }) =>
		hasBorder ? `0.8rem solid ${color}` : `none`};
	color: ${({ color }) => color};
	display: flex;
	align-items: center;
	justify-content: space-between;
	font-size: 1.4rem;
	padding: 1rem;
	max-height: 6.8rem;
	margin: ${({ margin }) => margin};

	&.dismissed {
		opacity: 0;
		max-height: 0;
		overflow: hidden;
		transition: all 0.2s ease-in-out;
		padding: 0;
	}

	button {
		border: none;
		background: transparent;
		font-size: 2.4rem;
		color: ${({ color }) => color};
		padding: 0;
		line-height: 0;
		margin-right: 1.4rem;
	}
`;
