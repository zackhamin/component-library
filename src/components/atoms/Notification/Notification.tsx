import classNames from 'classnames';
import { colors } from 'Styles';
import { StyledNotification } from './Notification.styles';
import { forwardRef, ReactNode, useState } from 'react';

export interface NotificationProps {
	children: ReactNode;
	color?: string;
	backgroundColor?: string;
	hasBorder?: boolean;
	isDismissible?: boolean;
	margin?: string;
	rest?: any;
}

export const Notification = forwardRef<HTMLDivElement, NotificationProps>(
	(
		{
			children,
			color = colors.emerald,
			backgroundColor = colors.white,
			hasBorder = true,
			isDismissible = false,
			margin,
			...rest
		},
		ref
	) => {
		const [isDismissed, setIsDismissed] = useState(false);

		return (
			<StyledNotification
				color={color}
				backgroundColor={backgroundColor}
				hasBorder={hasBorder}
				data-testid="notification"
				className={classNames({ dismissed: isDismissed })}
				aria-hidden={isDismissed}
				margin={margin}
				ref={ref}
				{...rest}
			>
				{children}
				{isDismissible && (
					<button
						aria-label="Dismiss notification"
						onClick={() => setIsDismissed(true)}
						data-testid="dismiss-button"
					>
						&times;
					</button>
				)}
			</StyledNotification>
		);
	}
);
