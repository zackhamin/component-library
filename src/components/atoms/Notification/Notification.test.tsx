import { fireEvent, render } from '@testing-library/react';

import { Notification } from '../Notification';

describe('Notification', () => {
	const defaultProps = {
		color: '#000',
		backgroundColor: '#fff',
		hasBorder: true,
		isDismissible: true,
	};

	it('renders correctly with a dismiss button', () => {
		const { container } = render(
			<Notification {...defaultProps}>
				<span>This is a notification</span>
			</Notification>
		);
		expect(container).toMatchSnapshot();
	});

	it('renders correctly without a dismiss button', () => {
		const { container } = render(
			<Notification {...defaultProps} isDismissible={false}>
				<span>This is a non dismissable notification</span>
			</Notification>
		);
		expect(container).toMatchSnapshot();
	});

	it('removes the notification from the DOM when the dismiss button is clicked', () => {
		const { getByTestId } = render(
			<Notification {...defaultProps}>
				<span>This is a notification</span>
			</Notification>
		);

		const notificationComponent = getByTestId('notification');
		const dismissButton = getByTestId('dismiss-button');
		fireEvent.click(dismissButton);
		expect(notificationComponent).not.toBeVisible();
	});
});
