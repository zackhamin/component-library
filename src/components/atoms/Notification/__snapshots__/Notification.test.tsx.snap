// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification renders correctly with a dismiss button 1`] = `
.c0 {
  background: #fff;
  border-left: 0.8rem solid #000;
  color: #000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-size: 1.4rem;
  padding: 1rem;
  max-height: 6.8rem;
}

.c0.dismissed {
  opacity: 0;
  max-height: 0;
  overflow: hidden;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  padding: 0;
}

.c0 button {
  border: none;
  background: transparent;
  font-size: 2.4rem;
  color: #000;
  padding: 0;
  line-height: 0;
  margin-right: 1.4rem;
}

<div>
  <div
    aria-hidden="false"
    class="c0"
    color="#000"
    data-testid="notification"
  >
    <span>
      This is a notification
    </span>
    <button
      aria-label="Dismiss notification"
      data-testid="dismiss-button"
    >
      Ã—
    </button>
  </div>
</div>
`;

exports[`Notification renders correctly without a dismiss button 1`] = `
.c0 {
  background: #fff;
  border-left: 0.8rem solid #000;
  color: #000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-size: 1.4rem;
  padding: 1rem;
  max-height: 6.8rem;
}

.c0.dismissed {
  opacity: 0;
  max-height: 0;
  overflow: hidden;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  padding: 0;
}

.c0 button {
  border: none;
  background: transparent;
  font-size: 2.4rem;
  color: #000;
  padding: 0;
  line-height: 0;
  margin-right: 1.4rem;
}

<div>
  <div
    aria-hidden="false"
    class="c0"
    color="#000"
    data-testid="notification"
  >
    <span>
      This is a non dismissable notification
    </span>
  </div>
</div>
`;
